File: payment-service/strategies/UPIProcessor.js
import { PaymentProcessor } from './PaymentProcessor.js';
import { v4 as uuidv4 } from 'uuid';

export class UPIProcessor extends PaymentProcessor {
    constructor() {
        super();
        this.PAYMENT_METHOD = 'UPI';
        this.PROCESSING_DELAY_MS = 2000; // UPI is faster
    }

    async process(transaction) {
        console.log(`Processing UPI payment for transaction: ${transaction.id}`);
        
        return new Promise((resolve) => {
            setTimeout(() => {
                // Simulate success/failure (90% success rate for UPI)
                const isSuccess = Math.random() < 0.9;
                
                if (isSuccess) {
                    const upiTransactionId = 'UPI' + uuidv4().substring(0, 8).toUpperCase();
                    console.log(`UPI payment successful for transaction: ${transaction.id} with UPI ID: ${upiTransactionId}`);
                    resolve(PaymentProcessor.success(upiTransactionId, 'UPI payment completed successfully'));
                } else {
                    const failureReason = this.getRandomUPIFailureReason();
                    console.warn(`UPI payment failed for transaction: ${transaction.id} - ${failureReason}`);
                    resolve(PaymentProcessor.failure(failureReason));
                }
            }, this.PROCESSING_DELAY_MS);
        });
    }

    getPaymentMethod() {
        return this.PAYMENT_METHOD;
    }

    canProcess(transaction) {
        return super.canProcess(transaction) && transaction.amount <= 100000; // UPI limit
    }

    getRandomUPIFailureReason() {
        const reasons = [
            'Insufficient balance',
            'UPI PIN incorrect',
            'Transaction timeout',
            'Bank server unavailable',
            'Daily limit exceeded'
        ];
        return reasons[Math.floor(Math.random() * reasons.length)];
    }
}

File: backend/src/main/java/com/upi/gateway/backend/strategy/impl/UPIProcessor.java
package com.upi.gateway.backend.strategy.impl;

import com.upi.gateway.backend.model.Transaction;
import com.upi.gateway.backend.strategy.PaymentProcessor;
import com.upi.gateway.backend.strategy.PaymentResult;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ThreadLocalRandom;

@Component
@Slf4j
public class UPIProcessor implements PaymentProcessor {
    
    private static final String PAYMENT_METHOD = "UPI";
    private static final int PROCESSING_DELAY_MS = 2000; // UPI is faster
    
    @Override
    public CompletableFuture<PaymentResult> process(Transaction transaction) {
        log.info("Processing UPI payment for transaction: {}", transaction.getId());
        
        return CompletableFuture.supplyAsync(() -> {
            try {
                // Simulate UPI processing time
                Thread.sleep(PROCESSING_DELAY_MS);
                
                // Simulate success/failure (90% success rate for UPI)
                boolean isSuccess = ThreadLocalRandom.current().nextDouble() < 0.9;
                
                if (isSuccess) {
                    String upiTransactionId = "UPI" + UUID.randomUUID().toString().substring(0, 8).toUpperCase();
                    log.info("UPI payment successful for transaction: {} with UPI ID: {}", 
                            transaction.getId(), upiTransactionId);
                    return PaymentResult.success(upiTransactionId, "UPI payment completed successfully");
                } else {
                    String failureReason = getRandomUPIFailureReason();
                    log.warn("UPI payment failed for transaction: {} - {}", transaction.getId(), failureReason);
                    return PaymentResult.failure(failureReason);
                }
                
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                log.error("UPI processing interrupted for transaction: {}", transaction.getId());
                return PaymentResult.failure("Processing interrupted");
            }
        });
    }
    
    @Override
    public String getPaymentMethod() {
        return PAYMENT_METHOD;
    }
    
    @Override
    public boolean canProcess(Transaction transaction) {
        return PAYMENT_METHOD.equalsIgnoreCase(transaction.getPaymentMethod()) &&
               transaction.getAmount().doubleValue() <= 100000; // UPI limit
    }
    
    private String getRandomUPIFailureReason() {
        String[] reasons = {
            "Insufficient balance",
            "UPI PIN incorrect",
            "Transaction timeout",
            "Bank server unavailable",
            "Daily limit exceeded"
        };
        return reasons[ThreadLocalRandom.current().nextInt(reasons.length)];
    }
}